src/filesystem/drive/inner.rs

```diff
@@ -48,11 +62,75 @@ impl InnerDrive {
             ))
     }
 
+    fn mark_ancestors_dirty(&mut self, node_id: NodeId) -> Result<(), OperationError> {
+        // Changes have happened in `node_id` walk up its parents to root marking nodes as dirty
+        let mut node_id = node_id;
+        self.dirty_nodes.push(node_id);
+        while let Some(parent_perm_id) = self.by_id(node_id)?.parent_id() {
+            // Setting a node's parent ID to its own ID is done in some operations such as moving a node
+            // to temporarily orphan it. Obviously this would cause an infinite loop and needs to be caught
+            if parent_perm_id == self.by_id(node_id)?.permanent_id() {
+                break;
+            }
+            node_id = self.lookup_internal_id(&parent_perm_id)?;
+            self.dirty_nodes.push(node_id);
+        }
+        Ok(())
+    }
+
+    pub(crate) async fn clean_drive(&mut self) {
+        // Not being able to find a Node should not necessarily be an error
+        // Should this take PermanentId? Can the node Id shift out from underneath me?
+
+        // Take the dirty node list (replacing it with an empty Vec)
+        // Reverse the list and then remove duplicates
+        // (can't use Vec::dedup since that only removes consecutive duplicates)
+        let mut dirty = std::mem::take(&mut self.dirty_nodes);
+        dirty.reverse();
+        let mut node_list: Vec<_> = dirty.iter().unique().collect();
```

Comment:
A HashSet will solve the de-duplication issue here (is this is the only case where we're making use of IterTools? that would allow us to drop that dependency as well). We'd loose the ordering but...

I'm not sure the reverse here is valuable. Things are added by walking from a leaf to a parent but if multiple operations occur the parent make be present in the list already before its child and will get hit / incorrectly calculated before the child does.

Solving that ordering issue is a bit tricky. If the dirty list was a HashSet I would probably convert it to a mutable Vec at this point and pop things off it just like you're doing now. If it has children, and any of its children are still present in the list of dirty ones it could be added to the end of the list delaying its encoding until after its children. This might be easier to do if the dirty list tracks PermanentIds instead as the child references could directly be checked against the HashSet

Reply:
Yup I get what you're doing now. I wasn't thinking in terms of reversed paths here. Make sense and I appreciate the diagram hahah

Reply:
I just pushed a commit that removes the `itertools` dependency. In doing so the logic in this section of code changed some, I have convinced myself it is equivalent but would love a second set of eyes.